### ISSUE TITLE
Ignore invalid repositories (NotGitRepository: No git repository was found at <path>)

### ISSUE THREAD

**Issue by NaoPross** [jonashaag/klaus#176]
Hello, I was trying to use Klaus under Alpine Linux but the server crashes because there are other files in the `KLAUS_REPOS_ROOT` path, and I couldn't remove them since they are autogenerated by Finder from Mac OS X (the folder is a samba share on a Mac mounted under Linux with cifs).

Right now I've bodged a solution by modifying `klaus/__init__.py` like so to ignore everything but `*.git` directories but maybe it would be better to implement something like a list of ignored files or to catch the exception.
```
@@ -17,7 +17,7 @@
 
     def __init__(self, repo_paths, site_name, use_smarthttp, ctags_policy='none'):
         """(See `make_app` for parameter descriptions.)"""
-        repo_objs = [FancyRepo(path) for path in repo_paths]
+        repo_objs = [FancyRepo(path) for path in repo_paths if path.endswith(".git")]
         self.repos = dict((repo.name, repo) for repo in repo_objs)
         self.site_name = site_name
         self.use_smarthttp = use_smarthttp

```

Traceback:
```
alpinevm:~# /etc/local.d/gitwebserver.start 
+ gunicorn --env KLAUS_SITE_NAME=Git Server --env KLAUS_REPOS_ROOT=/srv/git/repos -b 0.0.0.0:80 klaus.contrib.wsgi_autoreload
[2017-03-12 21:48:41 +0000] [2197] [INFO] Starting gunicorn 19.6.0
[2017-03-12 21:48:41 +0000] [2197] [INFO] Listening at: http://0.0.0.0:80 (2197)
[2017-03-12 21:48:41 +0000] [2197] [INFO] Using worker: sync
[2017-03-12 21:48:41 +0000] [2200] [INFO] Booting worker with pid: 2200
Reloading repository list...
[2017-03-12 21:48:42 +0000] [2200] [ERROR] Error handling request /
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 135, in handle
    self.handle_request(listener, req, client, addr)
  File "/usr/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 176, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/lib/python2.7/site-packages/klaus/contrib/wsgi_autoreload.py", line 44, in app
    *args, **kwargs
  File "/usr/lib/python2.7/site-packages/klaus/__init__.py", line 121, in make_app
    ctags_policy,
  File "/usr/lib/python2.7/site-packages/klaus/__init__.py", line 20, in __init__
    repo_objs = [FancyRepo(path) for path in repo_paths]
  File "/usr/lib/python2.7/site-packages/dulwich/repo.py", line 694, in __init__
    "No git repository was found at %(path)s" % dict(path=root)
NotGitRepository: No git repository was found at /srv/git/repos/.DS_Store
```

**Comment by jonashaag** [comment]
Hm not sure about this. If we exclude non-git directories by default then the user doesn't get any warning about broken repositories and may be left wondering why their repositories don't show up. If we hard-code the list of repositories at start-time, similar to how `KLAUS_REPOS` works, then the server needs to be restarted each time a new repo is added. We might simply ignore `.DS_Store` and other autogenerated files.

Not sure...

**Comment by NaoPross** [comment]
Yeah, but isn't it kinda a standard to have `.git` at the end of a bare repository?
Otherwise another way to solve the problem could be by checking if `path` is a directory like this:
```repo_objs = [FancyRepo(path) for path in repo_paths if not os.path.isfile(path)]```

**Comment by jonashaag** [comment]
Thing is you can also put up non-bare repos with klaus. Simply run it from your repo checkout with `klaus .`. That's the way I mostly use it.

I feel like the directory check is no better than the explicit `.DS_Store` exclusion... hmm

**Comment by xNephe** [comment]
Hello,

I just had a similar issue (but with the `.ssh` folder of my git user).

I guess a correct solution could be:
- ignore all hidden folder (.*) in `wsgi_autoreloading`. A git repository is not hidden, hidden folders should be totally ignored.
- handle the error in `klaus.__init__` and issue a log. If a user wonder why he does not see a repos, the log will give him the answer and it will be more understandable than an internal error. It would also allow to access the other repositories when one is broken, and in my opinion, a partial service is better than no service at all.

Also, it is a common pattern to have hidden folder alongside git repositories (bare repositories are often in the home of a git user), so it would be nice to have this situation handled!

**Comment by jonashaag** [comment]
> * handle the error in `klaus.__init__` and issue a log.

The `__init__` method of a class is no place for any kind of logging. (except maybe deprecation warnings)

> a partial service is better than no service at all.

Disagree on this if it means you're potentially missing out on errors. Better to fail immediately and in an obvious way than to attempt to continue operation with errors (hello, PHP!)

**Comment by jonashaag** [comment]
Fixed with #240
